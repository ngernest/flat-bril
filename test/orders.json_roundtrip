{
  "functions": [
    {
      "args": [
        {
          "name": "a",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "zero",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "a",
            "zero"
          ],
          "dest": "is_neg",
          "funcs": [],
          "labels": [],
          "op": "lt",
          "type": "bool"
        },
        {
          "args": [
            "is_neg"
          ],
          "funcs": [],
          "labels": [
            "mul_neg_one",
            "abs_res"
          ],
          "op": "br"
        },
        {
          "label": "mul_neg_one"
        },
        {
          "dest": "neg_one",
          "op": "const",
          "type": "int",
          "value": -1
        },
        {
          "args": [
            "a",
            "neg_one"
          ],
          "dest": "a",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "label": "abs_res"
        },
        {
          "args": [
            "a"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "abs",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "a",
          "type": "int"
        },
        {
          "name": "b",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "a",
            "b"
          ],
          "dest": "q",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "b",
            "q"
          ],
          "dest": "aq",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "a",
            "aq"
          ],
          "dest": "mod",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "mod"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "mod",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "a",
          "type": "int"
        },
        {
          "name": "b",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "label": "while.cond"
        },
        {
          "args": [
            "a",
            "b"
          ],
          "dest": "mod",
          "funcs": [
            "mod"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "dest": "zero",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "mod",
            "zero"
          ],
          "dest": "is_term",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "is_term"
          ],
          "funcs": [],
          "labels": [
            "while.finish",
            "while.body"
          ],
          "op": "br"
        },
        {
          "label": "while.body"
        },
        {
          "args": [
            "b"
          ],
          "dest": "a",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "mod"
          ],
          "dest": "b",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "while.cond"
          ],
          "op": "jmp"
        },
        {
          "label": "while.finish"
        },
        {
          "args": [
            "b"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "gcd",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "a",
          "type": "int"
        },
        {
          "name": "b",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "zero",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "a",
            "zero"
          ],
          "dest": "a_is_zero",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "a_is_zero"
          ],
          "funcs": [],
          "labels": [
            "check_b",
            "is_good"
          ],
          "op": "br"
        },
        {
          "label": "check_b"
        },
        {
          "args": [
            "b",
            "zero"
          ],
          "dest": "b_is_zero",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "b_is_zero"
          ],
          "funcs": [],
          "labels": [
            "special_case",
            "is_good"
          ],
          "op": "br"
        },
        {
          "label": "special_case"
        },
        {
          "args": [
            "zero"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        },
        {
          "label": "is_good"
        },
        {
          "args": [
            "a",
            "b"
          ],
          "dest": "ab",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "ab"
          ],
          "dest": "ab",
          "funcs": [
            "abs"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "a",
            "b"
          ],
          "dest": "gcdab",
          "funcs": [
            "gcd"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "ab",
            "gcdab"
          ],
          "dest": "lcm",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "lcm"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "lcm",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "u",
          "type": "int"
        },
        {
          "name": "n",
          "type": "int"
        },
        {
          "name": "use_lcm",
          "type": "bool"
        }
      ],
      "instrs": [
        {
          "label": "for.cond"
        },
        {
          "args": [
            "u",
            "n"
          ],
          "dest": "is_term",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "is_term"
          ],
          "funcs": [],
          "labels": [
            "for.finish",
            "for.body"
          ],
          "op": "br"
        },
        {
          "label": "for.body"
        },
        {
          "args": [
            "use_lcm"
          ],
          "funcs": [],
          "labels": [
            "lcm",
            "gcd"
          ],
          "op": "br"
        },
        {
          "label": "lcm"
        },
        {
          "args": [
            "u",
            "n"
          ],
          "dest": "lcm",
          "funcs": [
            "lcm"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "lcm",
            "u"
          ],
          "dest": "ordu",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "for.body.print"
          ],
          "op": "jmp"
        },
        {
          "label": "gcd"
        },
        {
          "args": [
            "u",
            "n"
          ],
          "dest": "gcdun",
          "funcs": [
            "gcd"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "n",
            "gcdun"
          ],
          "dest": "ordu",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "label": "for.body.print"
        },
        {
          "args": [
            "u",
            "ordu"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "dest": "one",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "u",
            "one"
          ],
          "dest": "u",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "for.cond"
          ],
          "op": "jmp"
        },
        {
          "label": "for.finish"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "orders"
    },
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        },
        {
          "name": "use_lcm",
          "type": "bool"
        }
      ],
      "instrs": [
        {
          "dest": "zero",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "dest": "u",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "n"
          ],
          "dest": "n",
          "funcs": [
            "abs"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "zero",
            "u"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "args": [
            "u",
            "n",
            "use_lcm"
          ],
          "funcs": [
            "orders"
          ],
          "labels": [],
          "op": "call"
        }
      ],
      "name": "main"
    }
  ]
}
