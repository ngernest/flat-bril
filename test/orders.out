@abs
	op: const		dest: zero		type: int		value: 0	
	op: lt   		dest: is_neg		type: bool		args: ["a", "zero"]	
	op: br   		args: ["is_neg"]		labels: ["mul_neg_one", "abs_res"]
.mul_neg_one
	op: const		dest: neg_one		type: int		value: -1	
	op: mul  		dest: a 		type: int		args: ["a", "neg_one"]	
.abs_res
	op: ret  		args: ["a"]	
@mod
	op: div  		dest: q 		type: int		args: ["a", "b"]	
	op: mul  		dest: aq		type: int		args: ["b", "q"]	
	op: sub  		dest: mod		type: int		args: ["a", "aq"]	
	op: ret  		args: ["mod"]	
@gcd
.while.cond
	op: call 		dest: mod		type: int		args: ["a", "b"]		funcs: ["mod"]
	op: const		dest: zero		type: int		value: 0	
	op: eq   		dest: is_term		type: bool		args: ["mod", "zero"]	
	op: br   		args: ["is_term"]		labels: ["while.finish", "while.body"]
.while.body
	op: id   		dest: a 		type: int		args: ["b"]	
	op: id   		dest: b 		type: int		args: ["mod"]	
	op: jmp  		labels: ["while.cond"]
.while.finish
	op: ret  		args: ["b"]	
@lcm
	op: const		dest: zero		type: int		value: 0	
	op: eq   		dest: a_is_zero		type: bool		args: ["a", "zero"]	
	op: br   		args: ["a_is_zero"]		labels: ["check_b", "is_good"]
.check_b
	op: eq   		dest: b_is_zero		type: bool		args: ["b", "zero"]	
	op: br   		args: ["b_is_zero"]		labels: ["special_case", "is_good"]
.special_case
	op: ret  		args: ["zero"]	
.is_good
	op: mul  		dest: ab		type: int		args: ["a", "b"]	
	op: call 		dest: ab		type: int		args: ["ab"]		funcs: ["abs"]
	op: call 		dest: gcdab		type: int		args: ["a", "b"]		funcs: ["gcd"]
	op: div  		dest: lcm		type: int		args: ["ab", "gcdab"]	
	op: ret  		args: ["lcm"]	
@orders
.for.cond
	op: eq   		dest: is_term		type: bool		args: ["u", "n"]	
	op: br   		args: ["is_term"]		labels: ["for.finish", "for.body"]
.for.body
	op: br   		args: ["use_lcm"]		labels: ["lcm", "gcd"]
.lcm
	op: call 		dest: lcm		type: int		args: ["u", "n"]		funcs: ["lcm"]
	op: div  		dest: ordu		type: int		args: ["lcm", "u"]	
	op: jmp  		labels: ["for.body.print"]
.gcd
	op: call 		dest: gcdun		type: int		args: ["u", "n"]		funcs: ["gcd"]
	op: div  		dest: ordu		type: int		args: ["n", "gcdun"]	
.for.body.print
	op: print		args: ["u", "ordu"]	
	op: const		dest: one		type: int		value: 1	
	op: add  		dest: u 		type: int		args: ["u", "one"]	
	op: jmp  		labels: ["for.cond"]
.for.finish
	op: ret  	
@main
	op: const		dest: zero		type: int		value: 0	
	op: const		dest: u 		type: int		value: 1	
	op: call 		dest: n 		type: int		args: ["n"]		funcs: ["abs"]
	op: print		args: ["zero", "u"]	
	op: call 		args: ["u", "n", "use_lcm"]		funcs: ["orders"]
